### ISSUE TITLE
Switching to P010 colour format breaks AMD HW AV1

### ISSUE THREAD

**Issue by empZealoth** [obsproject/obs-studio#10193]
### Operating System Info

Windows 11

### OBS Studio Version

30.0.2

### OBS Studio Log URL

https://obsproject.com/logs/tSO2db1BOnxNJl4Q

### Expected Behavior

Succesfully record HDR footage using AMD HW AV1 by switching to P010 colour format and Rec. 2100 (PQ) colour space as well as switching capture colour space to Rec. 2100 (PQ)

### Current Behavior

Switching to P010 colour space causes recording to fail with weird/irrelevant error
![image](https://github.com/obsproject/obs-studio/assets/10992509/2ee11f53-bbcb-4842-b110-e6dc7067c641)
I already verified both my AMD drivers and OBS are up-to-date

The log line is also pretty cryptic:
12:23:05.451: [fallback-amf-av1] amf_av1_create_fallback: AMFComponent::Init failed: AMF_INVALID_ARG

### Steps to Reproduce

1. I'm using W11 and 7800XT GPU and these settings:
![image](https://github.com/obsproject/obs-studio/assets/10992509/9332b973-cb5b-4b75-8dea-eebdfc701d5b)
![image](https://github.com/obsproject/obs-studio/assets/10992509/4e95980d-6c06-46a1-8244-344da6adde26)
![image](https://github.com/obsproject/obs-studio/assets/10992509/8990a01b-111b-4062-b5ca-f49f63d108ae)
2. In my inexperienced/limited testing it's enough to switch to P010 colour format to get the error in my setup
3. Start Recording
4. Error -> recording cancelled

...

### Anything else we should know?

Current settings work perfectly when I'm using AMD HW AV1 with NV12 and Rec.709, so at least some part of my hardware encoder is clearly working

**Comment by SuslikV** [comment]
has related thread on the forum:
https://obsproject.com/forum/threads/amd-hw-av1-hdr-nonsensical-errors.172990/

**Comment by lextra2** [comment]
Have you tried recording HDR with HEVC? If so, does it work?
Trying to narrow down if the issue is codec specific here.

**Comment by MattyFresh68** [comment]
Can confirm at least on HEVC that 10 bit works when PreAnalysis is disabled
![image](https://github.com/obsproject/obs-studio/assets/66339760/89d600fc-1711-4a24-9edc-3b36906fa525)

**Comment by empZealoth** [comment]
> `QVBR`, `HQVBR` & `HQCBR` all set `EnablePreAnalysis=true` [But PreAnalysis only works with NV12](https://github.com/GPUOpen-LibrariesAndSDKs/AMF/blob/master/amf/doc/AMF_Video_PreAnalysis_API.md#12-overview)
> 
> Use Rate Control `CQP` if you want to record in HDR Your title is misleading

Thanks, that worked, in a sense of "I could record P010 successfully after switching to CQP"


How is it misleading?
My recording was working fine in NV12, switching to P010 breaks it with a nonsensical error that tells you nothing. Googling also returned zero useful information

Obviously OBS can't do anything about the fact that HQVBR requires preanalysis and that only works with NV12, but it sure can provide a more meaningful error

Basically your message minus the snark would explain everything perfectly

**Comment by Fenrirthviti** [comment]
@lextra2 Please keep comments on-topic, and don't berate users for not having knowledge you think they should posses.

As for this issue, I agree that if this is a known universal fail condition we should have some kind of protection or warning against it. @empZealoth Can you adjust the title to reflect the error is with P010 and specific rate control methods so this can be better tracked?

**Comment by RytoEX** [comment]
> Obviously OBS can't do anything about the fact that HQVBR requires preanalysis and that only works with NV12, but it sure can provide a more meaningful error

> As for this issue, I agree that if this is a known universal fail condition we should have some kind of protection or warning against it.

I'm not sure that we can reliably know that QVBR, HQVBR & HQCBR will forever set `EnablePreAnalysis=true`. That does seem to be true now, but I think that option is set internally by AMF. I'm not sure that AMF provides any extended error messages or if it only returns the error code (in this case, AMF_INVALID_ARG for "invalid argument" or "memory type or size are invalid").

We can probably add detection on our end for this (check color format and rate control method?), but we'll have to be mindful if AMF changes what it does or supports.

**Comment by empZealoth** [comment]
Can you not check for `EnablePreAnalysis=true` directly? Technically you should be able to enable it with other rate control methods manually too, right?

I changed the title. If it still needs to be changed just say what needs to be improved

**Comment by lextra2** [comment]
@RytoEX 
You could check for the selected Color Format and selected Rate Control, and then give the user an error if incompatibility arises.

Also, [AMD is probably working on making PreAnalysis work with P010](https://github.com/GPUOpen-LibrariesAndSDKs/AMF/issues/399#issuecomment-1664012956)

**Comment by RytoEX** [comment]
> Can you not check for `EnablePreAnalysis=true` directly? Technically you should be able to enable it with other rate control methods manually too, right?

I'm not sure the property itself is initialized in AMF until we try to call Init, since I don't think we are setting it ourselves. I'm actually not clear from the AMF docs if these rate control methods forcefully enable PreAnalysis, or if you must enable it for them to work. From the AMF docs:
> QVBR, HQVBR and HQCBR are only supported if PreAnalysis is enabled.


> You could check for the selected Color Format and selected Rate Control, and then give the user an error if incompatibility arises.
> 
> Also, [AMD is probably working on making PreAnalysis work with P010](https://github.com/GPUOpen-LibrariesAndSDKs/AMF/issues/399#issuecomment-1664012956)

Yes, I [mentioned this](https://github.com/obsproject/obs-studio/issues/10193#issuecomment-1925993487) as a possibility above, with the caveat that we will have to keep on top of what changes AMD continues to make:
> We can probably add detection on our end for this (check color format and rate control method?), but we'll have to be mindful if AMF changes what it does or supports.
